File: python/tests/parity/consumer.py
import asyncio
import signal
from datetime import datetime
from typing import Any

from bullmq.job import Job
from bullmq.worker import Worker
from tests.parity.utils import getParityBackendOptions, logEvent, readDefinitions


def createWorker(definition: dict[str, Any], backendOptions):
    async def processor(job: Job, _job_token):
        logEvent(
            "job-started",
            {
                "timestamp": int(datetime.now().timestamp() * 1000),
                "test_id": str(definition.get("id")),
                "job_name": job.name,
                "test_secret": job.data.get("test_secret"),
                "job_secret": job.data.get("job_secret"),
            },
        )

        sleepTime = definition.get("simulation", {}).get("sleep", 0)
        if sleepTime > 0:
            await asyncio.sleep(sleepTime / 1000)
        attemptsToFail = definition.get("simulation", {}).get("fail", 0)
        if job.attemptsMade < attemptsToFail:
            raise Exception("Throw error to simulate processing failure")

        logEvent(
            "job-completed",
            {
                "timestamp": int(datetime.now().timestamp() * 1000),
                "test_id": str(definition.get("id")),
                "job_name": job.name,
                "test_secret": job.data.get("test_secret"),
                "job_secret": job.data.get("job_secret"),
            },
        )

    worker = Worker(
        definition.get("id", ""),
        processor,
        dict(
            concurrency=definition.get("worker", {}).get("concurrency", 1),
            **backendOptions,
        ),
    )
    return worker


async def runConsumers():
    backendOptions = getParityBackendOptions()
    definitions = readDefinitions()
    workers: list[Worker] = []

    shutdown_event = asyncio.Event()

    def signal_handler(signal, frame):
        print(f"{signal} signal received, shutting down.")
        shutdown_event.set()

    signal.signal(signal.SIGTERM, signal_handler)
    signal.signal(signal.SIGINT, signal_handler)

    for definition in definitions:
        worker = createWorker(definition, backendOptions)
        workers.append(worker)

    # Ready event sent once all the workers are ready to accept jobs
    logEvent("ready")
    _ = await shutdown_event.wait()

    for worker in workers:
        await worker.close()


if __name__ == "__main__":
    asyncio.run(runConsumers())

File: rust/tests/parity/src/consumer.rs
use std::sync::Arc;
use tokio::{
    task::JoinSet,
    time::{sleep, Duration},
};

use bullmq::{
    options::RedisConnectionOptions,
    worker::{CancellationToken, ProcessorFn},
    Job, Worker, WorkerOptions,
};

use crate::{
    models::{JobData, ParityConsumerTestCase, ParityEventData},
    utils::{self, log_event},
};

async fn run_worker(port: u16, definition: Arc<ParityConsumerTestCase>) -> bullmq::Result<Worker> {
    let def_for_closure = Arc::clone(&definition);

    let processor: ProcessorFn = Arc::new(move |job: Job, _token: CancellationToken| {
        let definition = Arc::clone(&def_for_closure);
        Box::pin(async move {
            let job_data = serde_json::from_value::<JobData>(job.data().to_owned()).unwrap();

            log_event(
                "job-started".to_string(),
                ParityEventData::new_some(
                    definition.id.clone(),
                    job.name().to_string(),
                    job_data.clone(),
                ),
            );
            if definition.simulation.sleep > 0 {
                sleep(Duration::from_millis(definition.simulation.sleep)).await;
            }

            if job.attempts_made() < definition.simulation.fail {
                return Err(bullmq::Error::ProcessingError(
                    "Simulated failure".to_string(),
                ));
            }

            log_event(
                "job-completed".to_string(),
                ParityEventData::new_some(definition.id.clone(), job.name().to_string(), job_data),
            );

            // Check logic here
            Ok(serde_json::json!(null))
        })
    });

    let mut concurrency = 1;
    if definition.worker.concurrency > 1 {
        concurrency = definition.worker.concurrency;
    }

    let mut connection = RedisConnectionOptions::default();
    connection.url = format!("redis://localhost:{port}");

    let options = WorkerOptions::default()
        .concurrency(concurrency)
        .connection(connection);

    let worker = Worker::with_options(&definition.id, processor, options).await?;

    while let Some(_) = worker.next_event().await {
        // Do nothing, keeps the loop active
    }

    Ok(worker)
}

pub async fn setup_workers(port: u16) -> bullmq::Result<()> {
    let definitions = utils::read_consumer_definitions()?;
    let mut set = JoinSet::new();
    for definition in definitions {
        set.spawn(async move { run_worker(port, Arc::new(definition)).await });
    }

    log_event("ready".to_string(), None);

    set.join_all().await;

    Ok(())
}
